File: src/kotoba/lang/witness_quorum/orchestrator.clj
(ns kotoba.lang.witness-quorum.orchestrator
  "Witnessed write orchestrator.

  Composes the substrate-pure write primitive, the witness selector, and the
  quorum reducer into a single L1-witnessed write that returns only when:
    - the record was committed to PDS (via `write-fn`), AND
    - >= :quorum-threshold witness cells produced matching accept (or
      reject) attestations.

  CLJ port of this repo's original @etzhayyim/witness-quorum
  src/orchestrator.ts. JVM-only, and synchronous throughout (unlike the TS
  original's async/Promise plumbing) -- see the repo README for the
  adaptation notes below.

  Per kotoba-datomic SPEC S5 + ADR-2605231400 S\"Implementation plan\" #2-#4."
  (:require [kotoba.lang.witness-quorum.attestation :as attestation]
            [kotoba.lang.witness-quorum.quorum :as quorum]
            [kotoba.lang.witness-quorum.reputation :as reputation]
            [kotoba.lang.witness-quorum.selector :as selector]
            [kotoba.lang.witness-quorum.slashing :as slashing]
            [kotoba.lang.witness-quorum.stake :as stake])
  (:import [java.security MessageDigest]
           [java.util.concurrent ConcurrentHashMap LinkedBlockingQueue TimeUnit]
           [java.util.function Function]))

(defn write-with-witnesses
  "One-shot L1-witnessed write. Returns when the quorum state is decided
  (witnessed / rejected / escalated) or the timeout fires.

  `opts`:
    :write-fn        `(fn [write-opts] -> {:uri ... :cid ...})`. Adapts the
                      TS `WriteCapableClient.write` method into a plain
                      function -- idiomatic Clojure prefers a function over
                      a single-method object.
    :write-opts      passed through to `write-fn` (e.g. {:collection ... :record ... :rkey ...}).
    :fleet           seq of fleet-cell maps (see `selector/fleet-cell`).
    :rule            the membrane rule (provides default :quorum-size/:quorum-threshold/:escalation-policy).
    :transport       {:request-attestation (fn [req] ...)
                       :subscribe-attestations (fn [quorum-group] -> poll-fn)}
                      -- see `create-in-memory-witness-transport` below. This
                      adapts the TS `WitnessTransport` interface's
                      `subscribeAttestations` (which returned an
                      AsyncIterable) into a function returning a `poll-fn`
                      compatible with `quorum/collect-quorum`.
    :quorum-options  override; default derived from `rule`.
    :timeout-ms      total wall-clock budget. Default 30000.

  Returns {:uri ... :cid ... :selected-witnesses [...] :state <quorum-state>}."
  [{:keys [write-fn write-opts fleet rule transport quorum-options timeout-ms]}]
  (let [quorum-options (or quorum-options
                            {:quorum-size (:quorum-size rule)
                             :quorum-threshold (:quorum-threshold rule)
                             :escalation-policy (or (:escalation-policy rule) :council)})
        receipt (write-fn write-opts)
        selected (selector/select-witnesses (:cid receipt) fleet (:quorum-size quorum-options (:quorum-size rule)))
        qg (selector/quorum-group (:cid receipt))
        poll-fn ((:subscribe-attestations transport) qg)]
    (doseq [cell selected]
      ((:request-attestation transport) {:cell cell
                                          :record-uri (:uri receipt)
                                          :record-cid (:cid receipt)
                                          :record (:record write-opts)
                                          :rule rule}))
    (let [state (quorum/collect-quorum selected poll-fn (assoc quorum-options :timeout-ms (or timeout-ms 30000)))]
      {:uri (:uri receipt)
       :cid (:cid receipt)
       :selected-witnesses selected
       :state state})))

(defn write-with-witnesses-precommit
  "Pre-commit variant of `write-with-witnesses` (ADR-2607110300 Phase 2):
  the write is proposed to witnesses BEFORE it is applied, and `commit-fn`
  only runs if the quorum reaches `:witnessed`. This is the difference
  between post-hoc cosign (the record already exists when witnesses see it)
  and pre-commit quorum (witnesses gate whether the record is applied at
  all).

  `opts` differs from `write-with-witnesses` in two keys:
    :propose-fn  `(fn [write-opts] -> {:uri ... :cid ...})`. Computes the
                 uri/cid the record WOULD have -- e.g. a local content-hash
                 derivation -- WITHOUT persisting anything. Must be
                 deterministic: the same `write-opts` must yield the same
                 :cid `commit-fn` will later persist, since witnesses sign
                 against this proposed cid.
    :commit-fn   `(fn [write-opts receipt] -> commit-result)`. Called ONLY
                 when the quorum state's :kind is :witnessed. Performs the
                 actual write/append. Not called on :rejected, :escalated,
                 or :pending.
  All other opts (:write-opts :fleet :rule :transport :quorum-options
  :timeout-ms) are unchanged from `write-with-witnesses`.

  Returns {:uri ... :cid ... :selected-witnesses [...] :state <quorum-state>
           :committed? bool :commit-result (present only when :committed? true)}.

  Note (honesty, per ADR-2607110300): this is crash-fault tolerance among
  witnesses operated by a single party, not Byzantine consensus -- there is
  no fork-choice/view-change for conflicting concurrent proposals. Callers
  must not represent this as BFT."
  [{:keys [propose-fn commit-fn write-opts fleet rule transport quorum-options timeout-ms]}]
  (let [quorum-options (or quorum-options
                            {:quorum-size (:quorum-size rule)
                             :quorum-threshold (:quorum-threshold rule)
                             :escalation-policy (or (:escalation-policy rule) :council)})
        receipt (propose-fn write-opts)
        selected (selector/select-witnesses (:cid receipt) fleet (:quorum-size quorum-options (:quorum-size rule)))
        qg (selector/quorum-group (:cid receipt))
        poll-fn ((:subscribe-attestations transport) qg)]
    (doseq [cell selected]
      ((:request-attestation transport) {:cell cell
                                          :record-uri (:uri receipt)
                                          :record-cid (:cid receipt)
                                          :record (:record write-opts)
                                          :rule rule}))
    (let [state (quorum/collect-quorum selected poll-fn (assoc quorum-options :timeout-ms (or timeout-ms 30000)))
          witnessed? (= :witnessed (:kind state))]
      (cond-> {:uri (:uri receipt)
               :cid (:cid receipt)
               :selected-witnesses selected
               :state state
               :committed? witnessed?}
        witnessed? (assoc :commit-result (commit-fn write-opts receipt))))))

(defn write-with-witnesses-precommit-and-slash
  "write-with-witnesses-precommit, plus automatically applying the
  resulting quorum state to reputation/stake via
  slashing/apply-quorum-outcome (ADR-2607110300 Phase 4). Before this,
  reputation.clj and stake.clj were independently tested but nothing
  ever called them from the actual precommit flow -- a completed quorum
  round had no economic consequence for the witnesses who disagreed
  with it. This closes that gap: use this instead of
  write-with-witnesses-precommit when you want reputation/stake tracked
  automatically; use the plain version when you're managing those
  separately (e.g. batching many rounds before updating).

  Additional opts beyond write-with-witnesses-precommit:
    :reputation-db  current reputation db. Default reputation/empty-reputation.
    :stake-ledger   current stake ledger. Default stake/empty-ledger.
    :slash-amount   bond units removed from a disagreeing witness per
                    round. Default 10.

  Returns write-with-witnesses-precommit's result map, plus
  `:reputation-db'` / `:stake-ledger'` / `:slashed` (the UPDATED
  reputation-db, updated stake-ledger, and this round's slash audit
  trail) merged in. `:pending`/`:escalated` quorum states are NOT
  applied (not a decision -- matches
  reputation/record-quorum-outcomes's own caveat); in that case
  `:reputation-db'`/`:stake-ledger'` pass through unchanged and
  `:slashed` is empty."
  [{:keys [reputation-db stake-ledger slash-amount]
    :or {reputation-db reputation/empty-reputation
         stake-ledger stake/empty-ledger
         slash-amount 10}
    :as opts}]
  (let [result (write-with-witnesses-precommit opts)
        decided? (contains? #{:witnessed :rejected} (:kind (:state result)))
        {:keys [reputation-db stake-ledger slashed]}
        (if decided?
          (slashing/apply-quorum-outcome reputation-db stake-ledger (:state result) slash-amount)
          {:reputation-db reputation-db :stake-ledger stake-ledger :slashed []})]
    (assoc result :reputation-db' reputation-db :stake-ledger' stake-ledger :slashed slashed)))

;; --- In-memory transport (testing + integration smoke) ------------------

(defn- computing-queue ^LinkedBlockingQueue [^ConcurrentHashMap queues key]
  (.computeIfAbsent queues key
                     (reify Function
                       (apply [_ _] (LinkedBlockingQueue.)))))

(defn create-in-memory-witness-transport
  "In-memory WitnessTransport: routes each request-attestation to the
  matching cell handler on a background thread (`future`), and makes the
  produced attestation available to `subscribe-attestations`'s poll-fn via a
  `LinkedBlockingQueue` per quorum-group.

  Unlike the TS original's lazy queue/waiters dance (needed there to avoid
  losing attestations that race ahead of a subscriber showing up), a
  `LinkedBlockingQueue` obtained via `computeIfAbsent` on the *same*
  quorum-group key from both the producer and the consumer side makes the
  ordering irrelevant -- whichever side arrives first creates the queue,
  and blocking put/poll do the rest.

  `opts`:
    :cell-handlers        map from fleet-cell `:key` to a handler function
                           `(fn [request] -> attestation-map-or-nil)`.
                           Unknown cells produce no attestation (simulates a
                           non-responsive cell).
    :attestation-delay-ms  optional delay before the handler's result is
                           queued. Default 0."
  [{:keys [cell-handlers attestation-delay-ms]}]
  (let [queues (ConcurrentHashMap.)]
    {:request-attestation
     (fn [req]
       (future
         (when (and attestation-delay-ms (pos? attestation-delay-ms))
           (Thread/sleep (long attestation-delay-ms)))
         (when-let [handler (get cell-handlers (:key (:cell req)))]
           (when-let [att (handler req)]
             (.put (computing-queue queues (:quorum-group att)) att))))
       nil)

     :subscribe-attestations
     (fn [quorum-group]
       (let [q (computing-queue queues quorum-group)]
         (fn [remaining-ms]
           (let [item (.poll q (long (max 0 remaining-ms)) TimeUnit/MILLISECONDS)]
             (if item
               {:status :value :value item}
               {:status :timeout})))))}))

;; --- Deterministic test signer --------------------------------------------

(defn make-deterministic-test-signer
  "Test-only signer: produces a deterministic 32-byte \"signature\" via
  sha256(canonical-bytes || cell-id). Byte-stable across re-runs, which
  makes integration tests reproducible. Replace with
  `kotoba.lang.witness-quorum.signer/make-ed25519-cell-signer` in production
  cells."
  [cell-id]
  (fn [^bytes canonical-bytes]
    (let [cell-bytes (.getBytes ^String cell-id "UTF-8")
          combined (byte-array (+ (alength canonical-bytes) (alength cell-bytes)))]
      (System/arraycopy canonical-bytes 0 combined 0 (alength canonical-bytes))
      (System/arraycopy cell-bytes 0 combined (alength canonical-bytes) (alength cell-bytes))
      (.digest (MessageDigest/getInstance "SHA-256") combined))))

(defn make-standard-cell-handler
  "Convenience: build a default cell handler that runs the standard
  membrane validation pipeline and produces a signed attestation.

  `opts`: {:cell ... :signer ... :validators ...} (see
  `kotoba.lang.witness-quorum.attestation/produce-attestation`)."
  [{:keys [cell signer validators]}]
  (fn [req]
    (attestation/produce-attestation
     {:record-uri (:record-uri req)
      :record-cid (:record-cid req)
      :record (:record req)
      :cell cell
      :rule (:rule req)
      :validators validators
      :signer signer})))

File: src/orchestrator.ts
/**
 * kotoba-datomic witnessed write orchestrator.
 *
 * Composes the substrate-pure write primitives (`Etzhayyim.write`, the
 * witness selector, and the quorum reducer) into a single L1-witnessed
 * write that exits the call site only when:
 *   - the record was committed to PDS, AND
 *   - ≥quorumThreshold witness cells produced matching `accept` (or
 *     `reject`) attestations.
 *
 * Per kotoba-datomic SPEC §5 + ADR-2605231400 §"Implementation plan" #2-#4.
 */

import { produceAttestation, type MembraneRule, type MembraneValidators, type CellSigner } from "./attestation.js";
import { collectQuorum, type Attestation, type QuorumOptions, type QuorumState } from "./quorum.js";
import { selectWitnesses, type FleetCell } from "./witness-selector.js";

/** Minimal Etzhayyim shape this module depends on. Defined here so the
 *  module is callable from anywhere that has a write-capable client,
 *  including tests with mock substrate. */
export interface WriteCapableClient {
  write(opts: {
    collection: string;
    record: Record<string, unknown>;
    rkey?: string;
  }): Promise<{ uri: string; cid: string }>;
}

/** Per-witness call from the orchestrator side: "please attest this record".
 *  Implementations route the request to the cell (HTTP / NATS / atproto
 *  cross-actor invoke). Fire-and-forget — attestation records arrive
 *  separately on the subscription channel. */
export interface WitnessRequest {
  cell: FleetCell;
  recordUri: string;
  recordCid: string;
  record: Record<string, unknown>;
  rule: MembraneRule;
}

/** Transport abstraction. Production uses PDS XRPC (`requestAttestation`
 *  hits a per-cell endpoint, `subscribeAttestations` reads the
 *  `com.etzhayyim.kotoba-datomic.attestation` firehose filtered by `quorumGroup`).
 *  Tests use the in-memory transport in `createInMemoryWitnessTransport`. */
export interface WitnessTransport {
  requestAttestation(req: WitnessRequest): Promise<void>;
  subscribeAttestations(quorumGroup: string): AsyncIterable<Attestation>;
}

export interface WriteWithWitnessesOpts {
  client: WriteCapableClient;
  writeOpts: {
    collection: string;
    record: Record<string, unknown>;
    rkey?: string;
  };
  fleet: readonly FleetCell[];
  rule: MembraneRule;
  transport: WitnessTransport;
  /** Default: { quorumSize: rule.quorumSize, quorumThreshold: rule.quorumThreshold,
   *  escalationPolicy: rule.escalationPolicy ?? "council" }. */
  quorumOptions?: QuorumOptions;
  /** Total wall-clock budget for quorum collection. Default 30s. */
  timeoutMs?: number;
}

export interface WriteWithWitnessesResult {
  uri: string;
  cid: string;
  selectedWitnesses: FleetCell[];
  state: QuorumState;
}

/** One-shot L1-witnessed write. Returns when the quorum state is decided
 *  (witnessed / rejected / escalated) or the timeout fires. Caller is
 *  responsible for handling non-witnessed terminal states. */
export async function writeWithWitnesses(opts: WriteWithWitnessesOpts): Promise<WriteWithWitnessesResult> {
  const { client, writeOpts, fleet, rule, transport } = opts;
  const quorumOptions: QuorumOptions = opts.quorumOptions ?? {
    quorumSize: rule.quorumSize,
    quorumThreshold: rule.quorumThreshold,
    escalationPolicy: rule.escalationPolicy ?? "council",
  };

  // 1. PDS write — get the URI + CID. Witness selection is keyed off CID.
  const receipt = await client.write(writeOpts);

  // 2. Deterministic witness selection. Identical (cid, fleet) → identical witnesses.
  const selected = await selectWitnesses(receipt.cid, fleet, quorumOptions.quorumSize ?? rule.quorumSize);

  // 3. Fan out attestation requests. Fire-and-forget per witness; the
  //    subscription below collects the resulting attestations.
  const requests = selected.map((cell) =>
    transport.requestAttestation({
      cell,
      recordUri: receipt.uri,
      recordCid: receipt.cid,
      record: writeOpts.record,
      rule,
    }),
  );
  // Subscribe BEFORE awaiting the request promises so we don't miss
  // attestations that race ahead of the dispatch returns. Implementations
  // typically subscribe via PDS firehose, which is broadcast and lossless.
  const { quorumGroup } = await import("./witness-selector.js");
  const qg = await quorumGroup(receipt.cid);
  const source = transport.subscribeAttestations(qg);

  // The request dispatches must complete before we count timeouts —
  // otherwise a slow transport could starve the witness pool.
  await Promise.all(requests);

  const state = await collectQuorum(selected, source, {
    ...quorumOptions,
    timeoutMs: opts.timeoutMs ?? 30_000,
  });

  return {
    uri: receipt.uri,
    cid: receipt.cid,
    selectedWitnesses: selected,
    state,
  };
}

// ─── In-memory transport (testing + integration smoke) ─────────────

/** Per-cell handler: given the witness request, decide whether to produce
 *  an attestation (and what verdict to issue). Returning `null` simulates
 *  a non-responsive cell. */
export type InMemoryCellHandler = (req: WitnessRequest) => Promise<Attestation | null>;

export interface InMemoryWitnessTransportOpts {
  /** Map from FleetCell.key (= `${node}::${cellId}`) to handler. Unknown
   *  cells produce no attestation (simulates cell-not-online). */
  cellHandlers: Map<string, InMemoryCellHandler>;
  /** Optional delay (ms) between requestAttestation returning and the
   *  attestation appearing on subscribeAttestations. Helps tests cover
   *  the "collect across time" path. Default 0. */
  attestationDelayMs?: number;
}

/** In-memory WitnessTransport: routes each requestAttestation to the
 *  matching cell handler, captures the produced attestation, and yields
 *  it on subscribeAttestations. The first subscriber drains the queue;
 *  subsequent subscribers see the in-flight pool. */
export function createInMemoryWitnessTransport(opts: InMemoryWitnessTransportOpts): WitnessTransport {
  const queues = new Map<string, Attestation[]>();
  const waiters = new Map<string, Array<(att: Attestation) => void>>();

  return {
    async requestAttestation(req: WitnessRequest): Promise<void> {
      const handler = opts.cellHandlers.get(req.cell.key);
      // Fire-and-forget but capture promise so test errors surface.
      void (async () => {
        if (opts.attestationDelayMs) {
          await new Promise((resolve) => setTimeout(resolve, opts.attestationDelayMs));
        }
        if (!handler) return;
        const att = await handler(req);
        if (!att) return;
        const w = waiters.get(att.quorumGroup);
        if (w && w.length > 0) {
          const cb = w.shift()!;
          if (w.length === 0) waiters.delete(att.quorumGroup);
          cb(att);
          return;
        }
        const q = queues.get(att.quorumGroup) ?? [];
        q.push(att);
        queues.set(att.quorumGroup, q);
      })();
    },
    subscribeAttestations(quorumGroup: string): AsyncIterable<Attestation> {
      return {
        [Symbol.asyncIterator]() {
          return {
            async next(): Promise<IteratorResult<Attestation>> {
              const q = queues.get(quorumGroup) ?? [];
              if (q.length > 0) {
                const value = q.shift()!;
                if (q.length === 0) queues.delete(quorumGroup);
                return { value, done: false };
              }
              const att = await new Promise<Attestation>((resolve) => {
                const list = waiters.get(quorumGroup) ?? [];
                list.push(resolve);
                waiters.set(quorumGroup, list);
              });
              return { value: att, done: false };
            },
          };
        },
      };
    },
  };
}

// ─── Deterministic test signer ─────────────────────────────────────

/** Test-only signer: produces a deterministic 32-byte "signature" via
 *  sha256(canonicalBytes || cellId). Replace with libsignal / Ed25519 in
 *  production cells. */
export function makeDeterministicTestSigner(cellId: string): CellSigner {
  return async (canonicalBytes: Uint8Array): Promise<Uint8Array> => {
    const cellBytes = new TextEncoder().encode(cellId);
    const combined = new Uint8Array(canonicalBytes.length + cellBytes.length);
    combined.set(canonicalBytes, 0);
    combined.set(cellBytes, canonicalBytes.length);
    const digest = await crypto.subtle.digest("SHA-256", combined);
    return new Uint8Array(digest);
  };
}

/** Convenience: build a default cell handler that runs the standard
 *  membrane validation pipeline and produces a signed attestation. */
export function makeStandardCellHandler(args: {
  cell: FleetCell;
  signer: CellSigner;
  validators?: MembraneValidators;
}): InMemoryCellHandler {
  return async (req: WitnessRequest): Promise<Attestation | null> => {
    return produceAttestation({
      recordUri: req.recordUri,
      recordCid: req.recordCid,
      record: req.record,
      cell: args.cell,
      rule: req.rule,
      validators: args.validators,
      signer: args.signer,
    });
  };
}
